lib/puma/launcher.rb

```diff
@@ -48,18 +49,28 @@ def initialize(conf, launcher_args={})
 
       env = launcher_args.delete(:env) || ENV
 
-      @config.options[:log_writer] = @log_writer
+      @config.load
+      @config.clamp
+      @options = @config.options
+
+      @options[:log_writer] = @log_writer
+      @options[:logger] = @log_writer if clustered?
+      if @options[:bind_to_activated_sockets]
+        @options[:binds] = @binder.synthesize_binds_from_activated_fs(
+          @options[:binds],
+          @options[:bind_to_activated_sockets] == 'only'
+        )
+      end
 
       # Advertise the Configuration
       Puma.cli_config = @config if defined?(Puma.cli_config)
+      log_config if env['PUMA_LOG_CONFIG']
 
-      @config.load
-
-      @binder        = Binder.new(@log_writer, conf)
-      @binder.create_inherited_fds(ENV).each { |k| ENV.delete k }
-      @binder.create_activated_fds(ENV).each { |k| ENV.delete k }
+      @binder        = Binder.new(@log_writer, @options)
+      @binder.create_inherited_fds(env).each { |k| env.delete k }
+      @binder.create_activated_fds(env).each { |k| env.delete k }
```

Comment:
The only time `env` is passed into the launcher is [here](https://github.com/puma/puma/blob/ca201ef69757f8830b636251b0af7a51270eb68a/lib/puma/cli.rb#L66), which happens to be `ENV` as well, which is probably why it wasn't caught that here we don't use the local var bad pass in the constant instead.